lib/test/cdo/redcarpet/test_inline.rb

```diff
@@ -0,0 +1,47 @@
+require_relative '../../test_helper'
+require 'cdo/redcarpet/inline'
+
+class RedcarpetInlineRendererTest < Minitest::Test
+  def setup
+    @renderer = Redcarpet::Markdown.new(Redcarpet::Render::Inline)
+  end
+
+  def test_base
+    assert_equal "base string", @renderer.render("base string")
+  end
+
+  def test_inline_markdown
+    expected = "a <strong>string</strong> with <em>some</em> <a href=\"markdown\">basic</a>"
+    markdown = "a **string** with _some_ [basic](markdown)"
+    assert_equal expected, @renderer.render(markdown)
+  end
+
+  def test_paragraph
+    expected = "<strong>some</strong> <em>basic</em> <a href=\"markdown\">inline</a>and <strong>yet more</strong> markdown in a <em>second</em> paragraph"
+    markdown = "**some** _basic_ [inline](markdown)\n\nand **yet more** markdown in a _second_ paragraph"
```

Comment:
I find this mildly surprising.  I've got a potentially bad assumption about newlines and paragraph breaks - you may need to correct my mental model here.

It looks like your input contains whitespace (specifically, two newlines, or an ignored `paragraph` token) between two words, and output that does not:

| Markdown | Output | As rendered by the browser |
| --- | --- | --- |
| `[inline](markdown)\n\nand` | `<a href=\"markdown\">inline</a>and` | [inline](#example)and |

This isn't what I expected; I thought the renderer would preserve the newlines, which seems slightly better because it would lean on the browser's handling of whitespace, and not collapse the adjacent words into one.

| Markdown | Output | As rendered by the browser |
| --- | --- | --- |
| `[inline](markdown)\n\nand` | `<a href=\"markdown\">inline</a>\n\nand` | [inline](#example) and |

In fact, this seems more like what's happening with lists in a later test.  It's also the behavior I expect from a single newline in markdown (although I think GFM treats a single newline as a line break).

| Markdown | Output | As rendered by the browser |
| --- | --- | --- |
| `[inline](markdown)\nand` | `<a href=\"markdown\">inline</a>\nand` | [inline](#example) and |

(Or as rendered by http://fiddlesalad.com/markdown/, and I have no idea what flavor it's using...)
![image](https://user-images.githubusercontent.com/1615761/71942997-ee770900-3173-11ea-97ab-5f7a37de73e6.png)

This whole thought may be beside the point, since we're not likely to see block elements in translations for inline strings anyway; just wondering if you had thoughts on this.

Reply:
To be a little more specific, I'm surprised that this test passes...

```rb
  def test_simple_newline
    expected = "left\nright"
    markdown = "left\nright"
    assert_equal expected, @renderer.render(markdown)
  end
```

...while this one fails...

```rb
  def test_simple_paragraph
    expected = "left\n\nright"
    markdown = "left\n\nright"
    assert_equal expected, @renderer.render(markdown)
  end
```

Reply:
ah, that's a good point. This behavior could be easily updated by doing something like

```ruby
def paragraph(text)
  text + "\n\n"
end
```

but that would mean we'd also be appending newlines to the end of our content:

| Markdown | Output |
| --- | --- |
| `[inline](markdown)\n\nand` | `<a href=\"markdown\">inline</a>\n\nand\n\n` |

which means the test would have to be something like

```ruby
  def test_simple_paragraph
    expected = "left\n\nright\n\n"
    markdown = "left\n\nright"
    assert_equal expected, @renderer.render(markdown)
  end
```

I don't have a strong feeling either way about which direction is better; what do you think?

Reply:
to clarify: this is happening because the parser is still detecting and parsing the paragraphs, it's just no longer rendering `p` tags for them and is instead just rendering the raw text. Unlike with remark, I don't see a way in redcarpet to turn off that parsing logic.

Reply:
Interesting.  It's annoying that either way this is [different from the inline remark behavior](https://github.com/code-dot-org/code-dot-org/pull/32531/files#diff-4e663954966dc82fe4d934c6b1dd9663R35-R49).

If those are our two options, I think I prefer the existing behavior because it's least surprising/most correct when the user does the most common thing and writes "inline" markdown with no newlines:

| Markdown | Mode | Output |
| --- | --- | --- |
| `What a _pain_.` | Ignore paragraphs | `What a <em>pain</em>.` |
| `What a _pain_.` | Append newlines to paragraphs | `What a <em>pain</em>.\n\n` |

Even though I think it's a worse user experience when the user does the _less_ common thing and tries to use blocks in inline markdown.  Let's stick with the current experience.  Thanks!

Reply:
>  It's annoying that either way this is different from the inline remark behavior.

Agreed. Note that right now, the behavior for other syntaxes is also different; for example, a header `# foo` will be rendered as `foo` without the syntax intact. This is again because Redcarpet still requires us to parse and render the syntaxes and we haven't done the work to teach it how to render them back to the original syntax.

Hopefully we'll be able to eventually switch everything over to using remark, and resolve that discrepancy by simply getting rid of this renderer.